tests/unit/lms/db/_bulk_action_test.py

```diff
@@ -58,6 +77,60 @@ def test_upsert(self, db_session):
     def test_upsert_does_nothing_if_given_an_empty_list_of_values(self, db_session):
         assert BulkAction(db_session).upsert(self.TableWithBulkUpsert, []) == []
 
+    def test_upsert_with_onupdate_columns(self, db_session):
+        db_session.add_all(
+            [
+                self.TableWithBulkUpsertOnUpdate(
+                    id=1, name="existing", scalar=0, callable=0, sql=0, default=0
+                ),
+                self.TableWithBulkUpsertOnUpdate(
+                    id=2, name="existing", scalar=1, callable=1, sql=1, default=1
+                ),
+            ]
+        )
+        db_session.flush()
+
+        BulkAction(db_session).upsert(
+            self.TableWithBulkUpsertOnUpdate,
+            [
+                {"id": 1, "name": "update_existing"},
+            ],
+        )
+
+        rows = list(
+            db_session.query(self.TableWithBulkUpsertOnUpdate).order_by(
+                self.TableWithBulkUpsertOnUpdate.id.asc()
+            )
+        )
```

Comment:
I don't think you need to have both a sort and `Any.iterable.containing` as they provide the same benefit of being order independant.

As this little snippet of code is the same between the two tests and might make for a nice helper like:

```python
    def assert_has_rows(self, db_session, *attrs):
        rows = list(db_session.query(self.TableWithBulkUpsert))

        assert (
            rows
            == Any.iterable.containing(
                [
                    Any.instance_of(self.TableWithBulkUpsert).with_attrs(expected)
                    for expected in attrs
                ]
            ).only()
        )
```

Along with the suggested parameterisation above, this would make the end of the test quite snappy:
```python
        self.assert_has_rows(
            db_session,
            # 42 is the onupdate default value
            {"id": 1, "name": "update_existing", column: 42},
            {"id": 2, "name": "pre_existing_2", column: 1},
        )
```